File: LeetCode_Algorithm/173(Binary Search Tree Iterator)/173.js
var BSTIterator = function(root) {
  // sol1 T : 246ms 20% S : 55.6MB 62%
  this.arr = []; queue = []
  while (true) {
    if (root) {
      queue.push(root)
      root = root.left
    } else if (queue.length) {
      root = queue.pop()
      this.arr.push(root)
      root = root.right
    } else {
      break
    }
  }
  return null
};

BSTIterator.prototype.next = function() {
    if (this.arr.length) {
    let node = this.arr.shift()
    return node.val
    } else {
        return null
    }
};

BSTIterator.prototype.hasNext = function() {
  return this.arr.length > 0 ? true : false
};

File: LeetCode_Algorithm/173(Binary Search Tree Iterator)/173.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
'''
LeetCode 173 : Binary Search Tree Iterator
주어진 tree를 inorder로 순회하는 BSTIterator 만들기

sol1 T : 116ms 75% S : 20.6MB 12%
T : O(n) S : O(n)
트리가 주어졌을 때 init에서 트리를 inorder로 순회한 뒤 그 순회값을 모두 저장해
하나씩 출력한다.
'''
from typing import Optional
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        from collections import deque
        self.queue = deque([])
        self.arr = deque([])
        while True:
            if root:
                self.queue.append(root)
                root = root.left
            elif self.queue:
                node = self.queue.pop()
                self.arr.append(node)
                root = node.right
            else:
                break
        
    def next(self) -> int:
        node = self.arr.popleft()
        return node.val

    def hasNext(self) -> bool:
        return True if self.arr else False